crates/cxx-qt-lib/src/core/qurl.rs

```diff
@@ -63,21 +196,190 @@ pub struct QUrl {
+    pub fn set_path(&mut self, path: &ffi::QString) {
+        ffi::qurl_set_path(self, path)
+    }
+
+    /// Sets the query string of the URL to query.
+    pub fn set_query(&mut self, query: &ffi::QString) {
+        ffi::qurl_set_query(self, query)
+    }
+
+    /// Sets the scheme of the URL to scheme. As a scheme can only contain ASCII characters,
+    /// no conversion or decoding is done on the input. It must also start with an ASCII letter.
+    pub fn set_scheme(&mut self, scheme: &ffi::QString) {
+        ffi::qurl_set_scheme(self, scheme)
+    }
+
+    /// Parses url and sets this object to that value.
+    /// QUrl will automatically percent encode all characters that are not allowed in a URL
+    /// and decode the percent-encoded sequences that represent an unreserved character
+    /// (letters, digits, hyphens, underscores, dots and tildes).
+    /// All other characters are left in their original forms.
+    pub fn set_url(&mut self, url: &ffi::QString) {
+        ffi::qurl_set_url(self, url)
+    }
+
+    /// Sets the user info of the URL to userInfo.
+    pub fn set_user_info(&mut self, user_info: &ffi::QString) {
+        ffi::qurl_set_user_info(self, user_info)
+    }
+
+    /// Sets the URL's user name to userName.
+    pub fn set_user_name(&mut self, user_name: &ffi::QString) {
+        ffi::qurl_set_user_name(self, user_name)
+    }
+
+    /// Returns a human-displayable string representation of the URL.
+    /// The option RemovePassword is always enabled, since passwords should never be shown back to users.
+    pub fn to_display_string(&self) -> ffi::QString {
+        ffi::qurl_to_display_string(self)
+    }
+
+    /// Returns the encoded representation of the URL if it's valid; otherwise an empty QByteArray is returned.
+    pub fn to_encoded(&self) -> ffi::QByteArray {
+        ffi::qurl_to_encoded(self)
+    }
+
+    /// Returns an encoded copy of input. input is first converted to UTF-8,
+    /// and all ASCII-characters that are not in the unreserved group are percent encoded.
+    /// To prevent characters from being percent encoded pass them to exclude.
+    /// To force characters to be percent encoded pass them to include.
+    pub fn to_percent_encoding(
+        input: &ffi::QString,
+        exclude: &ffi::QByteArray,
+        include: &ffi::QByteArray,
+    ) -> ffi::QByteArray {
+        ffi::qurl_to_percent_encoding(input, exclude, include)
+    }
 
     /// Returns a QString representation of the URL.
-    pub fn to_qstring(&self) -> QString {
+    pub fn to_qstring(&self) -> ffi::QString {
         ffi::qurl_to_qstring(self)
     }
+
+    /// Returns the user info of the URL, or an empty string if the user info is undefined.
+    pub fn user_info_or_default(&self) -> ffi::QString {
+        ffi::qurl_user_info(self)
+    }
+
+    /// Returns the user name of the URL if it is defined; otherwise an empty string is returned.
+    pub fn user_name_or_default(&self) -> ffi::QString {
```

Comment:
Also missing `fn user_name(&self) -> Option<ffi::QString>`

Reply:
So i've not found yet in the spec ( https://datatracker.ietf.org/doc/html/rfc3986 ) if the user name can be empty or not...

Reply:
```
The user information, if present, is followed by a
   commercial at-sign ("@") that delimits it from the host.
```
It says "if present" about the user info.
To me that suggests the user may be non-existent.
Not sure whether that means the user name should be "empty" though..
So question is whether:
`https://@test.com` or `https://:@test.com` is a valid URL? :thinking:

Reply:
I'm trying different URL validators (mostly based on regexes) and they don't really come to a consensus.
Most accept URLs in the format `https://:@test.com` though :thinking: 
So I don't think we can assume empty usernames are invalid...
So like with password, let's stick with the _or_default variant for name.